File: backend/app/api/tasks.py
from flask import jsonify, request

from app.api import api
from app.api.plans import error_response
from app.extensions import db
from app.models import Plan, Task
from app.services.plans import ValidationError
from app.services.executions import CompletionConflict, complete_task, serialize_completion
from app.services.tasks import (
    create_task,
    delete_task,
    reopen_task,
    serialize_task,
    task_query,
    update_task,
)


def active_task(task_id: str):
    task = db.session.get(Task, task_id)
    return task if task and task.deleted_at is None else None


@api.get("/plans/<plan_id>/tasks")
def list_tasks(plan_id: str):
    if db.session.get(Plan, plan_id) is None:
        return error_response("계획을 찾을 수 없습니다.", status=404)
    try:
        statement = task_query(
            plan_id,
            query=request.args.get("q"),
            status=request.args.get("status"),
            priority=request.args.get("priority"),
            tag=request.args.get("tag"),
        )
    except ValidationError as exc:
        return error_response("할 일 조건이 올바르지 않습니다.", details=exc.errors)
    tasks = db.session.execute(statement).scalars()
    return jsonify(
        {
            "tasks": [serialize_task(task) for task in tasks],
            "sort": "priority → dueDate → createdAt → id",
        }
    )


@api.post("/plans/<plan_id>/tasks")
def post_task(plan_id: str):
    plan = db.session.get(Plan, plan_id)
    if plan is None:
        return error_response("계획을 찾을 수 없습니다.", status=404)
    try:
        task = create_task(plan, request.get_json(silent=True))
    except ValidationError as exc:
        return error_response("할 일을 저장할 수 없습니다.", details=exc.errors)
    return jsonify({"task": serialize_task(task)}), 201


@api.get("/tasks/<task_id>")
def get_task(task_id: str):
    task = active_task(task_id)
    if task is None:
        return error_response("할 일을 찾을 수 없습니다.", status=404)
    return jsonify({"task": serialize_task(task)})


@api.patch("/tasks/<task_id>")
def patch_task(task_id: str):
    task = active_task(task_id)
    if task is None:
        return error_response("할 일을 찾을 수 없습니다.", status=404)
    try:
        task = update_task(task, request.get_json(silent=True))
    except ValidationError as exc:
        return error_response("할 일을 고칠 수 없습니다.", details=exc.errors)
    return jsonify({"task": serialize_task(task)})


@api.post("/tasks/<task_id>/complete")
def post_complete_task(task_id: str):
    task = active_task(task_id)
    if task is None:
        return error_response("할 일을 찾을 수 없습니다.", status=404)
    try:
        task, event, replayed = complete_task(task, request.get_json(silent=True))
    except ValidationError as exc:
        return error_response("완료 요청이 올바르지 않습니다.", details=exc.errors)
    except CompletionConflict as exc:
        return error_response(str(exc), status=409)
    return jsonify({"task": serialize_task(task), "completionEvent": serialize_completion(event), "replayed": replayed})


@api.post("/tasks/<task_id>/reopen")
def post_reopen_task(task_id: str):
    task = active_task(task_id)
    if task is None:
        return error_response("할 일을 찾을 수 없습니다.", status=404)
    try:
        return jsonify({"task": serialize_task(reopen_task(task))})
    except CompletionConflict as exc:
        return error_response(str(exc), status=409)


@api.delete("/tasks/<task_id>")
def remove_task(task_id: str):
    task = active_task(task_id)
    if task is None:
        return error_response("할 일을 찾을 수 없습니다.", status=404)
    delete_task(task)
    return "", 204

File: frontend/src/api/tasks.ts
import type { Priority } from "./plans";

export type TaskStatus = "active" | "completed";

export interface Task {
  id: string;
  planId: string;
  content: string;
  status: TaskStatus;
  dueDate: string;
  priority: Priority;
  tags: string[];
  estimatedMinutes: number;
  durationUnit: "minutes";
  completedAt: string | null;
  deletedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface TaskInput {
  content: string;
  dueDate: string;
  priority: Priority;
  tags: string[];
  estimatedMinutes: number;
}

export interface TaskFilters {
  q?: string;
  status?: TaskStatus | "";
  priority?: Priority | "";
  tag?: string;
}

interface ApiErrorBody {
  error?: { message?: string; details?: Record<string, string> };
}

export async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    ...init,
    headers: { "Content-Type": "application/json", ...init?.headers },
  });
  if (response.status === 204) return undefined as T;
  const body = (await response.json()) as T & ApiErrorBody;
  if (!response.ok) {
    const detail = body.error?.details ? Object.values(body.error.details)[0] : undefined;
    throw new Error(detail ?? body.error?.message ?? "요청을 처리하지 못했습니다.");
  }
  return body;
}

export async function listTasks(planId: string, filters: TaskFilters): Promise<Task[]> {
  const params = new URLSearchParams();
  Object.entries(filters).forEach(([key, value]) => {
    if (value) params.set(key, value);
  });
  const suffix = params.size ? `?${params.toString()}` : "";
  return (await request<{ tasks: Task[] }>(`/api/plans/${planId}/tasks${suffix}`)).tasks;
}

export async function createTask(planId: string, input: TaskInput): Promise<Task> {
  return (
    await request<{ task: Task }>(`/api/plans/${planId}/tasks`, {
      method: "POST",
      body: JSON.stringify(input),
    })
  ).task;
}

export async function updateTask(id: string, input: Partial<TaskInput>): Promise<Task> {
  return (
    await request<{ task: Task }>(`/api/tasks/${id}`, {
      method: "PATCH",
      body: JSON.stringify(input),
    })
  ).task;
}

export async function completeTask(id: string, idempotencyKey: string): Promise<Task> {
  return (await request<{ task: Task }>(`/api/tasks/${id}/complete`, {
    method: "POST", body: JSON.stringify({ idempotencyKey }),
  })).task;
}

export async function reopenTask(id: string): Promise<Task> {
  return (await request<{ task: Task }>(`/api/tasks/${id}/reopen`, { method: "POST" })).task;
}

export async function deleteTask(id: string): Promise<void> {
  await request<void>(`/api/tasks/${id}`, { method: "DELETE" });
}

